### PR TITLE
KEP 06 - The next generation of Shipyard

### PR DESCRIPTION
**Pull request by johannes-b**

### CODE REVIEW

File: text/0006-shipyard-next-generation.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+version: 0.2.0
+kind: Shipyard
+metadata:
+  name: shipyard-abc
+spec:
+  stages:
+  - name: hardening
+    sequences:
+    - name: artifact-delivery
+      triggers:
```
Comment (warber) [inline text/0006-shipyard-next-generation.md:58]: @johannes-b without explicitly reading the spec, "triggers" can mean both, events that trigger the sequence OR events that are triggered by the sequence. Changing this to e.g., "triggered-by" would be more expressive.

File: text/0006-shipyard-next-generation.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+      - name: deployment
+        properties:
+          strategy: direct
+      - name: test
+        properties:
+          kind: functional
+      - name: evaluation 
+      - name: release
+
+  - name: "hardening"
```
Comment (warber) [inline text/0006-shipyard-next-generation.md:257]: @johannes-b the second stage is called "staging" in the old yaml

File: text/0006-shipyard-next-generation.md (hunk 1 above)
Comment (christian-kreuzberger-dtx) [reply text/0006-shipyard-next-generation.md:58]: I agree, if we only look at the lines 54 to 59, it looks like that artifact-delivery in stage hardening *triggers* a `dev.artifact-delivery.finished`, which does not make any sense.

Looking at the whole file though, it's clear to me that there are stages, a stage has sequences, and a sequence has *triggers*.

I don't like the looks of "triggered-by" in a YAML file, hence I would propose a more generic "on", e.g.:

```yaml
spec:
  stages:
  - name: hardening
    sequences:
    - name: artifact-delivery
      on:  
      - dev.artifact-delivery.finished
      tasks:
      - name: deployment
      - name: test
      - name: evaluation 
      ...
```

File: text/0006-shipyard-next-generation.md (hunk 2 above)
Comment (christian-kreuzberger-dtx) [reply text/0006-shipyard-next-generation.md:257]: good catch

Comment (johannes-b) [reply text/0006-shipyard-next-generation.md]: thx, fixed it.

File: text/0006-shipyard-next-generation.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
+# The next generation of Shipyard
+
+The next generation of Shipyard to allow custom delivery/remediation processes.
+
+## Motivation
+
+*There's always room for smart enhancements* - the Shipyard specification is no exception.
+
+A shipyard allows a **Site Reliability Engineer (SRE)** to specify WHAT should happen in the course of an artifact delivery or remediation process. Due to certain use cases, it has been identified that the current specification of the shipyard has limitations and hides details from the SRE. This KEP proposes an enhancement of the *Shipyard* specification to bring flexibility into the definition of the *WHAT*, but still provides an opinionated approach and smart defaults.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:9]: > A shipyard allows a **Site Reliability Engineer (SRE)**
As this is literally the intro to this topic, it sounds like Shipyard is something that will be primarily used by SREs.

I would remove the SRE from this introduction completely, as our end-users are not exclusively SREs, but also Developers, DevOps, etc....

> Due to certain use cases, it has been identified ...
This sounds like a weak motiviation, it basically accuses "certain use cases"...

Here is my take on this first paragraph:

The current specification of [Keptns Shipyard](https://github.com/keptn/spec/blob/0.1.7/shipyard.md) has several limitations and hides details from end-users. This KEP provides an enhancement of the *Shipyard* format such that it allows end-users to specify *WHAT* should happen in the course of an artifact delivery or remediation process. While it brings flexibility into the definition of the *WHAT*, it is still an opinionated approach with smart defaults.

File: text/0006-shipyard-next-generation.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+Please read this KEP in combination with the KEP: [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7); especially regarding the eventing part.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:11]: I would change this to:

In addition, please consider [KEP 07: Propose the concept of a Uniform] as it specifies who responds to certain events/tasks defined within the Shipyard.

File: text/0006-shipyard-next-generation.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+## Explanation
+
+:man:/:blonde_woman: The target persona of this KEP is an *SRE*, who is responsible for processes and workflows to automate application delivery or remediation operations.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:15]: > The target persona of this KEP is an *SRE*
change to:
For instance, one target persona of this KEP could be an *SRE (Site Reliability Engineer)*

File: text/0006-shipyard-next-generation.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+Please read this KEP in combination with the KEP: [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7); especially regarding the eventing part.
+
+## Explanation
+
+:man:/:blonde_woman: The target persona of this KEP is an *SRE*, who is responsible for processes and workflows to automate application delivery or remediation operations. 
+
+Before getting started, a brief recap of the important terms and definitions used in this KEP:
+
+- **Project stage:** A project stage (or just stage) defines a logical space in an environment, which has a dedicated purpose for an application in a continuous delivery process. The project stages are ordered, meaning that a stage has at least a previous and/or next stage.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:19]: > meaning that a stage has at least a previous and/or next stage. 

This is not entirely true (e.g., single-stage projects). 

> The project stages are ordered

This also doesn't hold with the new shipyard format. Stages are not ordered, they can be triggered in any order (or even in parallel).

File: text/0006-shipyard-next-generation.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+:man:/:blonde_woman: The target persona of this KEP is an *SRE*, who is responsible for processes and workflows to automate application delivery or remediation operations. 
+
+Before getting started, a brief recap of the important terms and definitions used in this KEP:
+
+- **Project stage:** A project stage (or just stage) defines a logical space in an environment, which has a dedicated purpose for an application in a continuous delivery process. The project stages are ordered, meaning that a stage has at least a previous and/or next stage. 
+
+- **Shipyard:** A shipyard is the declarative means to divide an environment into project stages and to specify task sequences for each stage.
+
+- **Sequence:** A sequence declares a set of tasks, which are triggered by an external event (aka. domain event, e.g.: artifact built → deploy artifact, a problem occurred → execute remediation, etc.). A sequence executes the tasks in the given order.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:23]: > A sequence executes the tasks in the given order. 

Within a sequence, tasks are executed in the given order.

File: text/0006-shipyard-next-generation.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+Before getting started, a brief recap of the important terms and definitions used in this KEP:
+
+- **Project stage:** A project stage (or just stage) defines a logical space in an environment, which has a dedicated purpose for an application in a continuous delivery process. The project stages are ordered, meaning that a stage has at least a previous and/or next stage. 
+
+- **Shipyard:** A shipyard is the declarative means to divide an environment into project stages and to specify task sequences for each stage.
+
+- **Sequence:** A sequence declares a set of tasks, which are triggered by an external event (aka. domain event, e.g.: artifact built → deploy artifact, a problem occurred → execute remediation, etc.). A sequence executes the tasks in the given order. 
+
+- **Task:** A task is the smallest executable unit of a sequence. A task is kicked-off by a *triggered* event and has a life cycle of a *start* and *finished* event.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:25]: *start*, *status.changed*, and *finished* event.

File: text/0006-shipyard-next-generation.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+  - The evaluation always happens after a test execution even though the SRE already knows that no meaningful metric values are available (e.g., functional tests don't provide performance data). 
+
+  - The promotion of a deployed/released artifact happens automatically with no possibility to intervene.
+
+- It is not possible to add a task to a sequence, e.g., a task that takes care of pulling approval from a person in charge. 
+
+- The trigger (domain event) of a task sequence can not be set. 
+
+*Solution:* A revision of the shipyard specification v0.1.2 addresses the above-mentioned problems and gives SREs more flexibility in declaring processes.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:39]: > v0.1.2 
Did you mean v0.2.0

Also, I wouldn't say that "a revision of the shipyard spec" is the solution.

Maybe we should phrase it as follows:

*Solution*: By enhancing the Shipyard specificiation (v0.2.0) as detailed below, the above-mentioned problems are addressed and more flexibility for declaring processes is provided.

File: text/0006-shipyard-next-generation.md (hunk 1 above)
Comment (christian-kreuzberger-dtx) [reply text/0006-shipyard-next-generation.md:58]: @johannes-b any update on this? Will be stay with `triggers`?

File: text/0006-shipyard-next-generation.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+      - name: evaluation 
+      - name: test
+        properties:
+          kind: performance
+      - name: evaluation
+      - name: release 
+```
+
+*Meta-data:*
+* **version**: The version of the shipyard specification.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:77]: e.g., spec.keptn.sh/0.2.0

File: text/0006-shipyard-next-generation.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+        properties:
+          kind: performance
+      - name: evaluation
+      - name: release 
+```
+
+*Meta-data:*
+* **version**: The version of the shipyard specification. 
+* **kind**: Is set to `Shipyard`.
+* **metadata:** Contains at least the property *name*, which declares a unique name for the shipyard.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:79]: How is the name unique? Are we checking that in our implementation? Does it always have to be the project name?

File: text/0006-shipyard-next-generation.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+* **name:** A unique name of the sequence
+* **triggers** (`optional`): An array of events that trigger the task sequence.
+* **tasks:** An array of tasks executed by the sequence in the declared order.
+
+*Definition of Task:*
+
+* **name**: A unique name of the task
+* **properties** *(optional)*: Task properties as individual `key:value` pairs. These properties precise the task and are consumed by the unit that executes the task.
+
+**Note:** Reserved key tasks are: **deployment**, **test**, **evaluate**, **release**, **remediation**
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:97]: add **approval**

File: text/0006-shipyard-next-generation.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+      selector:
+        matchLabels:
+          status: failed
+  - hardening.release.finished:
+      selector:
+        matchExpressions:
+          - {key: status, operator: In, values: [warning, failed]}
+  tasks:
+  - name: rollback
+```
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:136]: This has not made it into the spec yet. Also, did we implement this or do we still need an issue for it?

File: text/0006-shipyard-next-generation.md [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+
+### Shipyard controller
+
+This KEP also proposes the implementation of a Keptn core component - the so-called *shipyard controller*. This controller implements the functionality described above. Besides, this controller is responsible for keeping sequences atomic. This means that a sequence is executed based on the shipyard, which was configured when the sequence was triggered. In other words, an update of a shipyard (and its containing sequences) does NOT impact ongoing sequences.
+
+### Keptn CloudEvents
+
+Based on the propsed concepts, changes on the Keptn CloudEvents are necessary.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:144]: typo: propsed -> proposed

File: text/0006-shipyard-next-generation.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+
+**(1) Reserved space in CloudEvent payload:**
+
+Before the shipyard controller sends out a `[task].triggered` event for a specific task, the controller adds a reserved space to the event payload. This reserved space is named after the task and can be consumed by the unit (Keptn-service) that executes the task. For example, a *deployment* task gets the reserved space of *deployment* in the payload (i.e., data field) of the CloudEvent. Per default, this reserved space is empty. Expect for tasks that have properties defined in the shipyard (e.g., `strategy: blue_green`) - for those tasks the properties are added to the reserved space:
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:148]: Typo? Expect for tasks  -> Except for tasks

File: text/0006-shipyard-next-generation.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+      "deployment": {
+        "strategy": "blue_green",
+        "traffic": "100",
+        "deploymentURI":"https://my-service.domain.com/" // Property added by Keptn-service
+      }
+    }
+  }
+  ```
+
+* The event selectors for Keptn-services, as proposed in [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7), work based on the information provided in this reserved space. For example:
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:179]: please add *KEP 07:* to the uniform link

File: text/0006-shipyard-next-generation.md [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+      }
+    }
+  }
+  ```
+
+
+
+## Breaking changes
+
+This KEP breaks the implementation of Keptn 0.6.0:
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:212]: should be 0.7.x now, right?

File: text/0006-shipyard-next-generation.md [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+    }
+  }
+  ```
+
+
+
+## Breaking changes
+
+This KEP breaks the implementation of Keptn 0.6.0: 
+* Keptn-services (aka. batteries) including contributed services need to be changed to react on a `[task].triggered`, and send a `[task].started` → `[task].finished`.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:213]: remove the *(aka. batteries)*, and changes this to "all Keptn-services including contributed services (keptn-contrib, keptn-sandbox)

File: text/0006-shipyard-next-generation.md [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+* Keptn CloudEvents must be changed.
+* A shipyard migration must take place as outlined below.
+
+### Upgrade path from Shipyard 0.1.2  to 0.2.0
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:217]: Let's not use 0.1.2, but *0.1.x* when we refer to the old Shipyard within this KEP

File: text/0006-shipyard-next-generation.md [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+
+The shipyard specification version 0.1.2 as used by Keptn 0.5.0 and 0.6.0 (last supported Keptn versions) has to be migrated to the new specification. Thus, the following spec:
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:219]: change "used by Keptn 0.5.0 and 0.6.0" to "used by Keptn versions 0.7.x and older"

remove the "(last supported Keptn versions)"

File: text/0006-shipyard-next-generation.md [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+          kind: real_user
+      - name: evaluation
+```
+
+## Ideas for enhancement
+
+- A task can have a property, which defines if the task should be executed depending on previous tasks: 
+  - `Only when all previous tasks have succeeded`
+  - `Even if a previous task has failed`
+  - `Only when a previous taks has failed`
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:309]: The following is just a discussion, and does not need to be addressed in the KEP!

I like it. GH Actions provides something similar:

* `if: ${{ always() }}` - https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#always
* `if: ${{ cancelled() }}` 
* `if: ${{ failure() }}`

In addition, in GH actions it is possible to refer to the result of previous steps, e.g. `${{ steps.test.outcome }}`, so another idea would be to only execute a task if the specified previous task has succeeded (or failed).

File: text/0006-shipyard-next-generation.md [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+
+## Open questions
+
+- Do we need a CloudEvent specification for sequence events?
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:313]: I would say yes, in a very basic form like
``sh.keptn.event.${stage}.${sequence}.(triggered|started|finished)``

File: text/0006-shipyard-next-generation.md (hunk 3 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:9]: Thanks for improving the motivation. I added a slight modification but I like your version better. I use it now. :)

File: text/0006-shipyard-next-generation.md (hunk 4 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:11]: I updated it accordingly.

File: text/0006-shipyard-next-generation.md (hunk 5 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:15]: Changed it accordingly. Thx!

File: text/0006-shipyard-next-generation.md (hunk 8 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:25]: *start* --> *started*
I forgot about the *status.changed* which is optional. 

I updated the KEP accordingly.

File: text/0006-shipyard-next-generation.md (hunk 9 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:39]: I re-phrased my sentence according to your suggestion and explicitly name the target version of the new Shipyard.

File: text/0006-shipyard-next-generation.md (hunk 12 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:97]: done

File: text/0006-shipyard-next-generation.md (hunk 14 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:144]: fixed

File: text/0006-shipyard-next-generation.md (hunk 15 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:148]: Is a typo. ;) Fixed it.

File: text/0006-shipyard-next-generation.md (hunk 16 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:179]: I added it and changed the link text.

File: text/0006-shipyard-next-generation.md (hunk 17 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:212]: Time passes by - back then it was 0.6.0. ;)

File: text/0006-shipyard-next-generation.md (hunk 18 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:213]: Removed the term *batteries*.

File: text/0006-shipyard-next-generation.md (hunk 6 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:19]: I gave it a try to improve this paragraph.

File: text/0006-shipyard-next-generation.md [hunk 23]
```diff
@@ -0,0 +1,315 @@
+# The next generation of Shipyard
+
+The next generation of Shipyard to allow custom delivery/remediation processes.
+
+## Motivation
+
+*There's always room for smart enhancements* - the Shipyard specification is no exception.
+
+The current specification of Keptn's [Shipyard](https://github.com/keptn/spec/blob/0.1.7/shipyard.md) has several limitations and hides details from the end-users. This KEP provides an enhancement of the *Shipyard* format such that it allows end-users to specify *WHAT* should happen in the course of an artifact delivery or remediation process. While it brings flexibility into the definition of the *WHAT*, it is still an opinionated approach with smart defaults.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:9]: *the end-users* - remove *the*

Other then that I like this paragraph much better now!

File: text/0006-shipyard-next-generation.md [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+
+In addition, please consider [KEP 07: Propose the concept of a Uniform](https://github.com/keptn/enhancement-proposals/pull/7) as it specifies which components responds to certain events/tasks defined within the Shipyard.
+
+## Explanation
+
+:man:/:blonde_woman: For instance, one target persona of this KEP could be an SRE (Site Reliability Engineer), who is responsible for processes and workflows to automate application delivery or remediation operations. 
+
+Before getting started, a brief recap of the important terms and definitions used in this KEP:
+
+- **Project stage:** A project stage (or just stage) defines a logical space in an environment, which has a dedicated purpose for an application in a continuous delivery process. At least one stage must be defined (e.g., for a single stage project), but multiple stages are allowed (e.g., for implementing a staging process using a development, hardening, and production stage). If multiple stages are defined, by default they are ordered according to the shipyard file. However, it is possible to change the default order by defining triggers mentioned below.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:19]: last part of this paragraph: add **as**
*by defining triggers **as** mentioned below*

File: text/0006-shipyard-next-generation.md [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+    * `hardening.artifact-delivery.started` 
+    * `hardening.artifact-delivery.finished` 
+
+1. Triggers: The array *triggers* contains all *domain events*(*) and sequence events with state finished to start this task sequence, but this array is optional. If it is not set, a sequence can be started with an event of type: `[stage.name].[sequence.name].triggered`. Note that only `<stage>.<task-sequence>.finished` events with `data.result` set to `pass` are able to trigger another task sequence.
+    * (*) Domain event: *An event that occurred in the business process, written in past tense; [see](https://en.wikipedia.org/wiki/Event_storming)*. This is fired by a human or tool to inform about a certain situation. For example, a `problem.open` event is fired by a monitoring tool when a service runs in a problem mode.
+
+1. Task events: For each task, a `[task].triggered` event is sent by the control plane. Those Keptn-services that have a subscription on this event, will react with a `[task].started` event, perform their functions, and finally confirm their execution with a `[task].finished`. E.g., for the *test* task, the following events occurre:
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:108]: typo: *occure* - remoev the *e*

File: text/0006-shipyard-next-generation.md [hunk 26]
```diff
@@ -0,0 +1,315 @@
...
+
+      }
+    }
+  }
+  ```
+
+## Breaking changes
+
+This KEP breaks the implementation of Keptn 0.7.0: 
+* All Keptn-services including contributed services (in the GitHub organizations: [keptn-contrib](https://github.com/keptn-contrib), [keptn-sandbox](https://github.com/keptn-sandbox)) need to be changed to react on a `[task].triggered` event, and send a `[task].started` → `[task].finished`.
```
Comment (christian-kreuzberger-dtx) [inline text/0006-shipyard-next-generation.md:211]: *in the GitHub organizations*

remove the *the*

File: text/0006-shipyard-next-generation.md (hunk 1 above)
Comment (christian-kreuzberger-dtx) [reply text/0006-shipyard-next-generation.md:58]: After talking to @jetzlstorfer we also thought about `triggeredByAny` (as the list of sequences is an array which is linked with an `OR` expression)

File: text/0006-shipyard-next-generation.md (hunk 25 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:108]: fixed it.

File: text/0006-shipyard-next-generation.md (hunk 26 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:211]: removed it.

File: text/0006-shipyard-next-generation.md (hunk 23 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:9]: thx, I removed the `the`

File: text/0006-shipyard-next-generation.md (hunk 11 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:79]: I checked out how Kubernetes is defining this field. I found: https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields

The definition says: `Data that helps uniquely identify the object, including a name string, UID, and optional namespace` 

I would also go with this definition and would not check it in our implementation.

File: text/0006-shipyard-next-generation.md (hunk 1 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:58]: We aggreed on: 
```
on:
  any:
    - sh.keptn.events
    - sh.keptn.events
  
  all:
    - sh.keptn.events
    - sh.keptn.events
```

Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:58]: We go with `on`, but will not implement the `any` / `all` property. By default, any applies.

File: text/0006-shipyard-next-generation.md (hunk 21 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:307]: We agreed on going with conditions on the sequence level and not on task-level.

File: text/0006-shipyard-next-generation.md (hunk 13 above)
Comment (johannes-b) [reply text/0006-shipyard-next-generation.md:136]: I filed an issue for that: https://github.com/keptn/keptn/issues/3028

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

**Pull request merged**